presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java

```diff
@@ -752,18 +789,79 @@ public IntArrayList computeEligibleRowIds(Page page)
+            if (inputKinds.size() > 1) {
+                // TODO Pass necessary pre-filled values to the delegate so it can evaluate these filters
+                throw new UnsupportedOperationException("Filters on a mix of partition and non-partition columns are not supported");
+            }
+
+            if (!evaluateFilterFunction(filterFunction)) {
+                filterOnPrefilledValuesFailed = true;
+                close();
+                break;
+            }
+        }
+
+        if (filterOnPrefilledValuesFailed) {
+            return true;
+        }
+
+        this.outputChannels = Arrays.copyOf(options.getOutputChannels(), options.getOutputChannels().length);
+
         // Remove non-regular and non-interim columns from internal and output channels
         int[] internalChannels = options.getInternalChannels();
         int[] outputChannels = options.getOutputChannels();
         for (int i = 0; i < internalChannels.length; i++) {
             ColumnMapping columnMapping = columnMappings.get(i);
             if (columnMapping.getKind() != REGULAR && columnMapping.getKind() != INTERIM) {
                 internalChannels[i] = -1;
-                outputChannels[i] = -1;
+                if (i < outputChannels.length) {
+                    outputChannels[i] = -1;
+                }
             }
         }
 
-        filterAndProjectPushedDown = delegate.pushdownFilterAndProjection(options);
+        filterAndProjectPushedDown = delegate.pushdownFilterAndProjection(
+                new PageSourceOptions(
+                        internalChannels,
+                        outputChannels,
+                        options.getReusePages(),
+                        remainingFilterFunctions.build().toArray(new FilterFunction[0]),
+                        options.getReorderFilters(),
+                        options.getTargetBytes(),
+                        options.getAriaFlags()));
         return filterAndProjectPushedDown;
     }
+
+    private Set<ColumnMappingKind> getInputKinds(FilterFunction filterFunction)
+    {
+        return Arrays.stream(filterFunction.getInputChannels())
+                .mapToObj(columnMappings::get)
+                .map(ColumnMapping::getKind)
+                .collect(toImmutableSet());
+    }
+
+    private boolean evaluateFilterFunction(FilterFunction filterFunction)
+    {
+        int[] inputChannels = filterFunction.getInputChannels();
+        Block[] blocks = new Block[inputChannels.length];
+        for (int i = 0; i < inputChannels.length; i++) {
+            blocks[i] = nativeValueToBlock(types[inputChannels[i]], prefilledValues[inputChannels[i]]);
+        }
+        PageSourceOptions.ErrorSet errorSet = new PageSourceOptions.ErrorSet();
+        int[] outputRows = new int[1];
+        int numRows = filterFunction.filter(new Page(blocks), outputRows, errorSet);
+        if (!errorSet.isEmpty()) {
+            filterOnPrefilledValuesException = Optional.of(errorSet.getFirstError(1));
```

Comment:
Since getFirstError can return null should this be `Optional.ofNullable(errorSet.getFirstError(1));`

Reply:
@elonazoulay getFirstError can return null only if error set is empty, which is not the case here.